pennylane/math/quantum.py

```diff
@@ -455,6 +455,71 @@ def reduced_dm(state, indices, check_state=False, c_dtype="complex128"):
     return density_matrix
 
 
+def purity(state, indices, check_state=False, c_dtype="complex128"):
+    r"""Computes the purity from a state vector or density matrix.
+
+    .. math::
+        \gamma = Tr(\rho^2)
+
+    where :math:`\rho` is the density matrix. The purity of a normalized quantum state satisfies
+    :math:`\frac{1}{d} \leq \gamma \leq 1`, where :math:`d` is the dimension of the Hilbert space.
+    A pure state has a :math:`\gamma` of 1.
+
+    It is possible to compute the purity of a sub-system from a given state. To find the purity of
+    the overall state, include all wires in the "indices" argument.
+
+    Args:
+        state (tensor_like): ``(2**N)`` state vector or ``(2**N, 2**N)`` density matrix.
+        indices (list(int)): List of indices in the considered subsystem.
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: purity of the considered subsystem.
+
+    ** Example **
+
+    >>> x = [1, 0, 0, 1] / np.sqrt(2)
+    >>> purity(x, [0, 1])
+    1.0
+    >>> purity(x, [0])
+    0.5
+
+    >>> x = [[1 / 2, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1 / 2]]
+    >>> purity(x, [0, 1])
+    0.5
+
+    .. seealso:: :func:`pennylane.qinfo.transforms.purity` and :func:`pennylane.purity`
+    """
+
+    density_matrix = reduced_dm(state, indices, check_state, c_dtype)
+    return _compute_purity(density_matrix)
+

```

Comment:
I'm worried about efficiency here. If we know the state is pure, then we can return 1 right away. This will avoid the expensive computation of the density matrix and the eigenvalues. On the other hand, if we have a density matrix already, then this is fine.

Can we split up the logic between pure states and mixed states? You can look at the other functions for examples, e.g. `fidelity`.

Reply:
I looked at the fidelity implementation, but the comments are somewhat misleading. It claims to check if the states are pure or mixed, while in fact it only checks if the state is a state vector or a density matrix, which makes sense, because a valid state vector is always a pure state, but when a density matrix is given, there is no way of knowing if it is pure or mixed without calculating the purity (or else there is no point calculating the purity). I can definitely do something similar, that is return 1 for state vectors, and compute the purity of density matrices.

Reply:
Right, I meant making the distinction between state vectors and density matrices, not pure and mixed states.